src/vm/hints/hint_utils.rs

```diff
@@ -13,29 +14,63 @@ use num_traits::{FromPrimitive, Signed, ToPrimitive, Zero};
 use std::collections::HashMap;
 use std::ops::{Neg, Shl, Shr};
 
+fn apply_ap_tracking_correction(
+    ap: MaybeRelocatable,
+    ref_ap_tracking: Option<&ApTracking>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<MaybeRelocatable, VirtualMachineError> {
+    if let (Some(ref_tracking_data), Some(hint_tracking_data)) = (ref_ap_tracking, hint_ap_tracking)
+    {
+        // check that both groups are the same
+        if ref_tracking_data.group != hint_tracking_data.group {
+            return Err(VirtualMachineError::InvalidTrackingGroup(
+                ref_tracking_data.group,
+                hint_tracking_data.group,
+            ));
+        }
+        if let MaybeRelocatable::RelocatableValue(relocatable) = ap {
+            let ap_diff = hint_tracking_data.offset - ref_tracking_data.offset;
+
+            Ok(MaybeRelocatable::from((
+                relocatable.segment_index,
+                relocatable.offset - ap_diff,
+            )))
+        } else {
+            Err(VirtualMachineError::InvalidApValue(ap))
```

Comment:
Is this an error or does revoking a reference (change in `group`) simply mean `ap` is a new address? We need to check this I think.

Reply:
this error raises if the `ap` register is not holding a memory address, which would be an unexpected error

Reply:
Aren't addresses the only thing `ap` (as register) can contain?